index.bs

```diff
@@ -261,16 +282,33 @@ dictionary MediaStreamTrackGeneratorInit {
     with |init|.{{MediaStreamTrackGeneratorInit/kind}}.
 4. Return |g|.
 
-### Attributes ### {#attributes-generator}
+### Attributes ### {#generator-attributes}
 <dl>
 <dt><dfn attribute for=MediaStreamTrackGenerator>writable</dfn></dt>
-<dd>Allows writing media frames the {{MediaStreamTrackGenerator}}, which is
-itself a {{MediaStreamTrack}}. If the {{MediaStreamTrack/kind}} attribute is `"audio"`,
-the stream will accept {{AudioData}} objects and fail with any other type. If
-{{MediaStreamTrack/kind}} is `"video"`, the stream will accept {{VideoFrame}} objects
-and fail with any other type. When a frame is written to {{MediaStreamTrackGenerator/writable}},
-the frame's `close()` method is automatically invoked, so that its internal resources are no longer
-accessible from JavaScript.
+<dd>Allows writing media frames to the {{MediaStreamTrackGenerator}}, which is
+itself a {{MediaStreamTrack}}. When this attribute is access for the first time,
+it MUST be initialized with the following steps:
+1. Initialize [=this=].{{MediaStreamTrackGenerator/writable}} to be a new {{WritableStream}}.
+2. <a dfn for="WritableStream">Set up</a> [=this=].{{MediaStreamTrackGenerator/writable}}, with its [=WritableStream/set up/writeAlgorithm=] set to [=writeFrame=] with [=this=] as parameter, [=WritableStream/set up/closeAlgorithm=] set to [=closeWritable=] with [=this=] as parameter, and [=WritableStream/set up/abortAlgorithm=] set to [=closeWritable=] with [=this=] as parameter.
+
+The <dfn>writeFrame</dfn> algorithm is given a |generator| and a |frame| as input. It is defined by running the following steps:
+1. If |generator|.{{MediaStreamTrack/kind}} equals `video` and |frame| is not a {{VideoFrame}} object, return [=a promise rejected with=] a {{TypeError}}.
+2. If |generator|.{{MediaStreamTrack/kind}} equals `audio` and |frame| is not an {{AudioData}} object, return [=a promise rejected with=] a {{TypeError}}.
+3. Send the media data backing |frame| to all live tracks connected to |generator|, possibly including |generator| itself.
+4. Invoke the `close` method of |frame|.
+5. Return  [=a promise resolved with=] undefined.
+
+The <dfn>closeWritable</dfn> algorithm is given a |generator| as input.
+It is defined by running the following steps.
+1. {{StopSource|Stop}} the generator as a MediaStreamTrack source.
```

Comment:
The link provides no information as to what should happen in this case, since this source is not "a device exposed by getUserMedia()."

Reply:
Surprisingly, I can't find any explicit mention that all tracks are ended when a source is stopped in w3.org/TR/mediacapture-streams/. I'm probably just missing something.

Reply:
I just linked to that "stopped" means. but you have a good point.
The MediaStream spec only defines a stopped source in the case  that all connected tracks end. But it does not have a concept of source being stopped for some other reason and that causing all tracks to end.

The closest definition for this is the ended event on MediaStreamTrack (see https://w3c.github.io/mediacapture-main/#event-mediastreamtrack-ended):
"The&nbsp;MediaStreamTrack&nbsp;object's source will no longer provide any data, either because the user revoked the permissions, or because the source device has been ejected, or because the remote peer permanently stopped sending data.""

So this, which would obviously correspond to the writable being closed, is not really called stopping a source, but just "the source will no longer provide any data", so updated the language to not refer to this as stopping a source, since that concept does not exist. I updated the algorithm to just end all connected tracks and refer to the algorithm to end tracks, which does exisst.